Fix CMR-related type hints

chuckwondo opened the pull request:
There were a number of type hints in `search.py` and `api.py` related to CMR queries that were incorrect. These were fixed. In addition, there were a number of other static type errors that were masked because of ignored `cmr` imports. Added type stubs for `python_cmr` library to unmask and address these additional type errors.

Limited static type changes as much as possible to only functions and methods dealing with CMR queries and results to keep this PR manageable.

Fixes #508

<!-- readthedocs-preview earthaccess start -->
----
📚 Documentation preview 📚: https://earthaccess--510.org.readthedocs.build/en/510/

<!-- readthedocs-preview earthaccess end -->

chuckwondo commented:
pre-commit.ci autofix

chuckwondo commented:
pre-commit.ci run

chuckwondo commented:
There seems to be some sort of discrepancy between pre-commit ci's ruff hook and locally running ruff because the pre-commit failure was due to import formatting that doesn't cause a problem locally. Specifically, pre-commit ci wanted to _remove_ a blank line on line 5 of `tests/unit/test_results.py`, but locally running ruff wants to _insert_ that blank line.

No idea why that's the case, but I'm concerned that this might cause a persistent build issue because now that I've pulled the autofix that I triggered above, and run `make lint`, I now get:

```plain
$ make lint
bash scripts/lint.sh
+ mypy earthaccess stubs tests
Success: no issues found in 31 source files
+ ruff check .
tests/unit/test_results.py:1:1: I001 [*] Import block is un-sorted or un-formatted
Found 1 error.
[*] 1 fixable with the `--fix` option.
```

That's _failing_ locally on the autofix that pre-commit ci applied because the autofix _removed_ a blank line. If I were to "fix" that locally and re-push, then pre-commit ci would again fail again because my local "fix" would _insert_ the blank line.

chuckwondo commented on tests/unit/test_results.py:
FYI, the changes in this file do not materially change the tests. This is simply a refactoring of the code to follow the convention used for integrating vcrpy with Unittest: https://vcrpy.readthedocs.io/en/latest/usage.html#unittest-integration

chuckwondo replied:
Also, this ends up renaming the cassettes, but the contents of the cassettes have not changed, although 1 cassette was added that didn't previously exist.

betolink reviewed (approved):
I think this PR cleans up and improves the typing used in earthaccess. The only observation/question I have is if from now on we'll have to write Python interfaces for new classes? (.pyi) Would we need some extra instructions on the "how to contribute" section? Python and typing doesn't feel intuitive for newcomers.

chuckwondo commented on earthaccess/typing_.py:
```diff
@@ -0,0 +1,49 @@
+    from typing import Dict, List, Mapping, Sequence, Tuple
+else:
+    from builtins import dict as Dict, list as List, tuple as Tuple
+    from collections.abc import Mapping, Sequence
```
Yes, I pulled all imports from `typing`, `typing_extensions`, and `collections.abc` into this module so that throughout the earthaccess code, we simply import from `.typing_` (or `earthaccess.typing_`) to avoid having to deal with all of the changes that have been made across Python versions.

There are various "collection" types prior to 3.9 that were exposed from the `typing` module, but as of 3.9, the recommendation is to import them from `collections.abc`. There's also the transition from using `List`, `Dict`, and `Tuple` to `list`, `dict`, and `tuple`, respectively, along with various other changes that are annoying to keep up with.

Therefore, this `typing_` module centralizes all such logic in one place, so throughout the rest of the code base, we can ignore it all and import all types from `typing_`.

chuckwondo commented on earthaccess/api.py:
```diff
@@ -78,9 +75,7 @@ def search_datasets(
-def search_data(
-    count: int = -1, **kwargs: Any
-) -> List[earthaccess.results.DataGranule]:
+def search_data(count: int = -1, **kwargs: Any) -> List[DataGranule]:
```
In this case, `Any` is the correct type. When typing `kwargs`, the type is the _value_ type. You cannot annotate the _key_ type for `kwargs` because it is a "special" dictionary where you can use _only_ strings (literal keywords) as keys, so you only annotate the _value_ type.

In this case, `Any` is the only reasonable thing we can do because there are no real restrictions on the types of values we can pass for various keywords. Unfortunately, Python does not provide a convenient means for us to fully type a large set of possible keyword arguments.

chuckwondo replied:
To clarify, in this particular case, we cannot reasonably narrow the value type, but that's not generally true. We might have a case for defining a more specialized function where we expect only values of a certain type(s) for keyword values.

chuckwondo commented on earthaccess/search.py:
```diff
@@ -644,46 +829,63 @@ def temporal(
 
-    def point(self, lon: str, lat: str) -> Type[GranuleQuery]:
+    @override
+    def point(self, lon: FloatLike, lat: FloatLike) -> Union[Self, Never]:
```
Yes, both are valid values, as `python_cmr` simply passes the values to the `float` function/constructor.

chuckwondo replied:
More specifically, the `FloatLike` type alias (`Union[str, SupportsFloat]`) indicates that we can pass either an `str` value (which could cause an exception if it's a string that cannot be parsed as a float), or any instance of a type that implements the `__float__` method, which the `float` function/constructor defers to.

chuckwondo commented on stubs/cmr/__init__.pyi:
I'm not sure what you mean. Do you mean do we have to define interfaces within `earthaccess`? Or do you mean do we have to include `.pyi` files? Either way, the answer is no. However, that does remind me that I forgot to add an empty `py.typed` file to this PR. By adding such a file, we're indicating that our code includes its own type hints, and there is no need for a separate type stub package.

chuckwondo commented on earthaccess/search.py:
```diff
@@ -160,14 +197,20 @@ def doi(self, doi: str) -> Type[CollectionQuery]:
         return self
 
-    def instrument(self, instrument: str) -> Type[CollectionQuery]:
+    def instrument(self, instrument: str) -> Union[Self, Never]:
```
The `Never` (or `NoReturn`, prior to Python 3.12) indicates that the function/method may throw an exception.

mfisher87 replied:
https://docs.python.org/3/library/typing.html#typing.Never

mfisher87 replied:
I'm not sure if I understand `Something | Never` annotation. My understanding is the reason to annotate `def foo() -> Never:` is to enable mypy to throw an error on `f = foo()`, since `foo` can't return. But if a function sometimes returns `Something` and sometimes returns `Never`, what additional checking does that enable?

My understanding is as a "bottom type", `Something | Never` is identical to `Something`!

I've read an argument that it serves to help document for humans that a possible outcome is an exception, but IMO that's a shortcoming of the type system that shouldn't be documented using the type system, since if you remove the possibility of an exception from the function and forget to also update the annotation, the annotation will be wrong and the typechecker won't know.

chuckwondo replied:
Just to clarify, a return type of `Something | Never` does not indicate that it is possible for a function to return `Never`. It indicates that the function possibly never returns at all (typically indicative of the possibility of raising an exception _instead_ of returning a value).

However, I agree that Python's type system has shortcomings, so using `Never` really only serves to help the reader, but in thinking more about this, and seeing that an IDE may not even show `Never` in the function's return type when it pops up info when hovering your mouse pointer over a function call (at least VS Code doesn't show it, perhaps because it is a bottom type), then it would seem that this perhaps is not even helpful to the reader (unless directly reading the source code where the hint is written).

Further, given that a return type of `Something | Never` can be very confusing, since in all other cases, a union type indicates the possible types of return values, including `Never` doesn't actually makes sense since you can _never_ return `Never`.

Therefore, I'm going to drop the use of `Never`, but I'll keep the `Raises` sections that I added to the docstrings, since that indeed does show up in the little mouse-over popups in IDEs, and is helpful to include in generated docs.

> if you remove the possibility of an exception from the function and forget to also update the annotation, the annotation will be "wrong" (no longer serving its intended purpose) and the typechecker won't know.

As an aside, this isn't a particularly strong argument for removing `Never` because this issue is not unique to `Never`. There are currently plenty of places in the current codebase where mypy doesn't complain, but should, due to incorrect return type hints. In fact, this PR corrected many of them that mypy wasn't complaining about because of the very loose configuration of mypy, so this issue is also a matter of making sure we're also configuring the type checker effectively.

I added a bit of strictness with this PR, but not too much. Only enough to help weed out some of the incorrect hints scoped within the context of this PR. I plan to incrementally increase strictness so that we don't have to touch too many parts of the code at once.

If we don't improve our type hints, then we may be doing more harm than good since incorrect type hints are perhaps worse than no type hints, as with documentation.

mfisher87 replied:
>Therefore, I'm going to drop the use of `Never`, but I'll keep the `Raises` sections that I added to the docstrings, since that indeed does show up in the little mouse-over popups in IDEs, and is helpful to include in generated docs.

That sounds great to me, thanks so much for looking in to IDE behavior around these annotations. These details of the developer experience aren't always paid attention to, and I love to see when they are! :heart: 


>I added a bit of strictness with this PR, but not too much. Only enough to help weed out some of the incorrect hints scoped within the context of this PR. I plan to incrementally increase strictness so that we don't have to touch too many parts of the code at once.
>
>If we don't improve our type hints, then we may be doing more harm than good since incorrect type hints are perhaps worse than no type hints, as with documentation.

:100: * :100: Couldn't agree more with the analysis and the approach. We need this and I really appreciate you're spearheading the work!

mfisher87 commented:
I want to say I'll review it this weekend, but I have some important paperwork due very soon that I've been postponing for far too long...

Let's see how that goes :)

chuckwondo commented:
> I want to say I'll review it this weekend, but I have some important paperwork due very soon that I've been postponing for far too long...
> 
> Let's see how that goes :)

I appreciate your critical eyes! 👀

chuckwondo commented:
pre-commit.ci autofix

mfisher87 commented:
:eyes:

mfisher87 commented on tests/unit/test_results.py:
```diff
@@ -58,132 +71,92 @@ def test_get_more_than_2000(self):
-
-            assert_unique_results(granules)
+        # Assert that we performed one 'hits' search and two 'results' search queries
+        self.assertEqual(len(self.cassette), 3)
```
Nice DRYing here!

mfisher87 reviewed (approved):
I think this looks great! One thing I'd like to see is some documentation about the purpose of and process for maintaining the `stubs/` directory. It's not immediately obvious to me that these are type stubs for `python_cmr`, because the import for that library is simply `cmr`, and my brain doesn't make that jump intuitively or immediately. It would also be good to document the long-term intention to upstream these instead of keeping them around!

Perhaps this could go in `CONTRIBUTING.md`. As I'm thinking that, I also think we've outgrown this single-file contributing doc, but that's a different issue (#521)

chuckwondo commented on earthaccess/typing_.py:
```diff
@@ -0,0 +1,46 @@
+"""
+Convenience module for importing types from the typing module, abstracting away
+the differences between Python versions.
```
The more I look at this (and also take into consideration your other comment on this file), I'm thinking that we should probably eliminate this file, and instead, as a matter of practice, always import from `typing_extensions`. That module is intentionally designed to smooth over Python version differences in the `typing` module. What say you?

mfisher87 replied:
I like that! :rocket:

betolink commented:
Should we merge this? @mfisher87 @chuckwondo

chuckwondo commented:
> I think this looks great! One thing I'd like to see is some documentation about the purpose of and process for maintaining the `stubs/` directory. It's not immediately obvious to me that these are type stubs for `python_cmr`, because the import for that library is simply `cmr`, and my brain doesn't make that jump intuitively or immediately. It would also be good to document the long-term intention to upstream these instead of keeping them around!
> 
> Perhaps this could go in `CONTRIBUTING.md`. As I'm thinking that, I also think we've outgrown this single-file contributing doc, but that's a different issue (#521)

I'm a bit torn on how we should handle this.

I started adding a section to `CONTRIBUTING.md` with details on how a contributor should either add a "sidecar" type stubs package, if available (e.g., `types-requests` for `requests`), or generate stubs (like I did in this PR for `python_cmr`), but the more I wrote, the more I thought that perhaps this is too much detail for adding to the guide.

It's starting to feel like we're including information that is documented elsewhere about dealing with type stubs in Python in general, and that we would just be repeating such general documentation (and perhaps not as well). It feels akin to including a section on "how to program in Python" because we use Python in this repo. A contributor should look elsewhere for such information.

If a contributor has difficulty resolving type check failures, they should reach out by asking a question in an issue, discussion, or PR, and one of us more familiar with type checking can assist. This should wash out during PR reviews, just like any other aspect of Python programming in general.

Regarding "promotion" of the included `python_cmr` stubs to either typeshed, or directly upstream to `python_cmr`, I suggest that perhaps we simply capture that in a new issue, rather than describing it here.

What say you?

chuckwondo commented:
> Should we merge this? @mfisher87 @chuckwondo

Not just yet. We're close, but there are a couple more questions open, which I posted in response to a couple of questions from @mfisher87

chuckwondo commented on scripts/lint.sh:
```diff
@@ -1,7 +1,6 @@
 
-set -e
-set -x
+set -ex
```
Mmm. Could do, but I'll pull the YAGNI card for now :-)

mfisher87 commented:
>It's starting to feel like we're including information that is documented elsewhere about dealing with type stubs in Python in general, and that we would just be repeating such general documentation (and perhaps not as well). It feels akin to including a section on "how to program in Python" because we use Python in this repo. A contributor should look elsewhere for such information.

I don't think we should duplicate information (as much as possible), more like a sentence or two that's relevant to our project's more curious contributors. Even if it's simply "We added type stubs for the untype python-cmr library, which we intend to eventually upstream. You shouldn't need to worry about this, but check here and here to learn more."

chuckwondo commented:
@mfisher87, I believe I've addressed everything now. The most recent tweaks:

- added a couple of sentences to the contributing guide on type stubs and the cmr stubs we have included
- removed the `typing_` module I had originally added to this PR with simply using `typing_extensions` (and also added it as a direct dependency, as we were relying on it as a transitive dep) -- I didn't globally make use of `typing_extensions` (or even look for other places where `typing` is used), but only used it in places where I had originally made use of the now-removed `typing_` module

mfisher87 commented on CONTRIBUTING.md:
```diff
@@ -84,6 +84,15 @@ Finally, for _development dependencies only_, you must add an entry to
+
+We have included type stubs for the untyped `python-cmr` library, which we
+intend to eventually upstream.  Since `python-cmr` exposes the `cmr` package,
+the stubs appear under `stubs/cmr`.
```
:100: to both new paragraphs, thanks!

mfisher87 reviewed (approved):
LGTM, thanks so much Chuck!! :heart_eyes:

chuckwondo commented:
Thanks @mfisher87! I'm willing to take on the responsibility of having merge privileges, if you'd like to grant them, so I can squash-merge my changes.

mfisher87 commented:
Invitation sent!
